core/arch/aarchxx/mangle.c

```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
+    for (i = 0; i < num_regs; i++)
+        usecount[i] = 0;
+    for (i = 0; i < num_regs; i++) {
+        if (regs[i] >= 0 && regs[i] != i)
+            ++usecount[regs[i]];
+    }
+
+    /* Set up register arguments that are not part of a cycle. */
+    do {
+        for (i = 0; i < num_regs; i++) {
+            if (regs[i] == i || usecount[i] != 0)
+                continue;
+            if (regs[i] == -1) {
+                insert_mov_immed_ptrsz(dcontext, opnd_get_immed_int(args[i]),
+                                       opnd_create_reg(regparms[i]),
+                                       ilist, instr, NULL, NULL);
+            } else if (regs[i] == -2 && opnd_get_reg(args[i]) == DR_REG_XSP) {
+                /* XXX: We could record which register has been set to the SP to
+                 * avoid repeating this load if several arguments are set to SP.
+                 */
+                insert_get_mcontext_base(dcontext, ilist, instr, regparms[i]);
+                PRE(ilist, instr, instr_create_restore_from_dc_via_reg
                     (dcontext, regparms[i], regparms[i], XSP_OFFSET));
-            } else if (opnd_get_reg(arg) != regparms[i]) {
-                POST(ilist, mark, XINST_CREATE_move(dcontext,
+            } else {
+                PRE(ilist, instr, XINST_CREATE_move(dcontext,
                                                     opnd_create_reg(regparms[i]),
-                                                    arg));
+                                                    args[i]));
+                if (regs[i] != -2)
+                    --usecount[regs[i]];
             }
-        } else {
-            /* FIXME i#1551, i#1569: we only implement naive parameter preparation,
-             * where args are all regs or immeds and do not conflict with param regs.
-             */
-            ASSERT_NOT_IMPLEMENTED(false);
+            regs[i] = i;
+            break;
+        }
+    } while (i < num_regs);
+
+    /* From now on it is safe to use LR as a temporary. */
+
+    /* Set up register arguments that are in cycles. */
+    for (;;) {
+        int first, tmp;
+        for (i = 0; i < num_regs; i++) {
+            if (regs[i] != i)
+                break;
+        }
+        if (i >= num_regs)
+            break;
+        first = i;
+        PRE(ilist, instr, XINST_CREATE_move(dcontext, opnd_create_reg(DR_REG_LR),
+                                            opnd_create_reg(regparms[i])));
+        do {
+            tmp = regs[i];
+            ASSERT(0 <= tmp && tmp < num_regs);
+            PRE(ilist, instr,
+                XINST_CREATE_move(dcontext, opnd_create_reg(regparms[i]),
+                                  tmp == first ? opnd_create_reg(DR_REG_LR) :
+                                  opnd_create_reg(regparms[tmp])));
+            regs[i] = i;
+            i = tmp;
+        } while (tmp != first);
+    }
+
+    /* Set up stack arguments that are (non-zero) constants or SP. */
```

Comment:
I've added a comment near the start of the function: /* The strategy here is to first set up ... (Yes, I do split infinitives!)